### ISSUE TITLE
Incorrect wrapping of proxy objects on every render

### ISSUE THREAD

**Issue by mfbx9da4** [realm/realm-js#4513]
### How frequently does the bug occur?

All the time

### Description

https://github.com/realm/realm-js/blob/master/packages/realm-react/src/useObject.tsx#L78

```
    // Wrap object in a proxy to update the reference on rerender ( should only rerender when something has changed )
    return new Proxy(object, {});
```

The intention of this code is that the reference will only change if the object has changed to play nicely with reacts immutability conventions. However, the hook is not invoked only when the hook fires. It is invoked anytime the parent component renders. The component could be rendering because a totally unrelated change to some other component state which will cause unnecessary Proxy wrapping, unnecessarily new references and break React's immutability conventions.

Easier to explain with code

```
const initialObject = {};
function useObject() {
  return new Proxy(initialObject, {});
}

export default function App() {
  const [count, setCount] = useState(0);
  const object = useObject();
  useEffect(() => {
    // This effect will be invoked everytime the component
    // renders (eg click the button), even though the object
    // never actually changed
    console.log("object changed", object);
  }, [object]);

  return (
    <div className="App">
      <h1>{count}</h1>
      <button onClick={() => setCount((x) => x + 1)}>Click me</button>
    </div>
  );
}
```

https://codesandbox.io/s/modern-field-7ety98?file=/src/App.js:69-642

### Can you reproduce the bug?

Yes, always

### Version

main

### What SDK flavour are you using?

Local Database only

### Are you using encryption?

No, not using encryption

### Platform OS and version(s)

ios

### Build environment

Which debugger for React Native: ..

**Comment by takameyer** [comment]
@mfbx9da4 Good catch.  Really appreciate the feedback.  We will investigate this as time allows and get back to you when we have something in review.

**Comment by takameyer** [comment]
@mfbx9da4 We have a good idea for fixing the proxy wrapper.  That being said, can you please make a separate issue for:

>Side issue: while the object cache and collection will make great strides to improving immutability, non-primitive values such as Dates, sets and dictionaries will not be immutably cached correctly. Are you aware of that?

We would also appreciate an example where this is not being handled correctly 🙂  Thanks so much for your feedback!

**Comment by mfbx9da4** [comment]
@takameyer what do you make of my singleton object proposal? Is there somewhere better to discuss that?

**Comment by takameyer** [comment]
@mfbx9da4 I think it's worth looking into.  We have discussed various ways we could improve or notification systems.  I think making a feature request would be a good way to invoke more discussion.

**Comment by sdandois** [comment]
Hi, I noticed that `useQuery` has the same problem. 

Is there a previous version where this works correctly? The README file caught my attention: 

> ## useQuery
> 
> Returns Realm.Results from a given type. This Hook will update on any changes to any Object in the Collection and return an empty array if the Collection is empty. The result of this can be consumed directly by the data argument of any React Native VirtualizedList or FlatList. If the component used for the list's renderItem prop is wrapped with React.Memo, then only the modified object will re-render.

```
const Component = () => {
  // ObjectClass is a class extending Realm.Object, which should have been provided in the Realm Config.
  // It is also possible to use the model's name as a string ( ex. "Object" ) if you are not using class based models.
  const collection = useQuery(ObjectClass);

  // The methods `sorted` and `filtered` should be wrapped in a useMemo.
  const sortedCollection = useMemo(() => collection.sorted(), [collection]);

  return (
    <FlatList data={sortedCollection} renderItem={({ item }) => <Object item={item}/>
  )
}
```

That is not working as expected, all items in my list are being re-rendered.

**Comment by sdandois** [comment]
As a workaround, I've implemented a caching hook on top of the ones provided by @realm/react:

```js
export const createRefCache = object => {
  return new Proxy(
    {},
    {
      get: function (target, property) {
        return object[property];
      },
      set: function (target, property, value) {
        object[property] = value;
      },
    },
  );
};

export const useCachedObject = (object, otherDeps = []) => {
  const [counter, setCounter] = useState(0);
  const forceRender = () => setCounter(n => n + 1);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const memoObject = useMemo(() => createRefCache(object), [counter, ...otherDeps]);

  useEffect(() => {
    const listener = (obj, changes) => {
      if (changes.deletions.length || changes.insertions.length || changes.modifications.length) {
        forceRender();
      }
    };
    memoObject.addListener(listener);

    return () => {
      memoObject.removeListener(listener);
    };
  }, [memoObject]);

  return memoObject;
};
```
Let me hear what you think about it!

**Comment by takameyer** [comment]
This has been fixed in `@realm/react@0.4.0`

**Comment by mordechaim** [comment]
@takameyer 
I'm not sure what I'm missing, but I clearly still observe the issue described in #4913 for both `useQuery` and `useObject`.

```
Android 10

"react-native": "^0.70.5",
"realm": "^11.2.0",
"@realm/react": "^0.4.1"
```

**Comment by dimonnwc3** [comment]
I still experience issue for both useQuery and useObject

**Comment by channeladam** [comment]
Argh, same. Just stumbled upon this issue after creating #5264 .
